robotframework-ls/src/robotframework_ls/impl/code_analysis.py

```diff
@@ -130,4 +161,28 @@ def collect_analysis_errors(completion_context):
             if len(errors) >= MAX_ERRORS:
                 # i.e.: Collect at most 100 errors
                 break
+
+    argument_names = list(ast_utils.iter_keyword_arguments_as_str(ast))
+    variable_assigns = list(ast_utils.iter_variable_assigns_as_str(ast))
+    variables_collector = _VariablesCollector()
+    collect_variables(completion_context, variables_collector)
+
+    for variable_usage_info in ast_utils.iter_variable_usage_tokens(ast):
```

Comment:
Another thing that occurred to me is that the scoping may not be correct here: in this implementation, when an argument is available for some keyword it seems it's made automatically available for all keywords for a match, which isn't correct. 

The implementation should probably be going into a Keyword and then collecting the variables which are valid just for that scope. Likewise, if a variable is used in a resource import, it needs to be defined globally and not inside some keyword import...